File: PA_75.py
#LC 208 - Implement Trie (Prefix Tree)
#Time Complexity - O(n)
#Space Complexity - O(n)
class TrieNode(object):
    
    def __init__(self):
        self.children = dict()
        self.endOfWord = False
        
class Trie(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()
        

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: None
        """
        currNode = self.root
        for i in range(len(word)):
            char = word[i]
            if char not in currNode.children:
                currNode.children[char] = TrieNode()
            currNode = currNode.children[char]
        currNode.endOfWord = True

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        currNode = self.root
        for i in range(len(word)):
            char = word[i]
            if char not in currNode.children:
                return False
            currNode = currNode.children[char]
        return currNode.endOfWord

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        currNode = self.root
        for i in range(len(prefix)):
            char = prefix[i]
            if char not in currNode.children:
                return False
            currNode = currNode.children[char]
        return True
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: PA75.java
//Leetcode: 208. Implement Trie (Prefix Tree)
//Time Complexity: for each method it is O(n) , n is the length of the word
//Space Complexity: O(m), where m is the no of nodes required to insert a new word. Each node is of fixed size 26. 
class Trie {
    class TrieNode{
        boolean isEnd;
        TrieNode[] children;
        TrieNode(){
            isEnd=false;
            children= new TrieNode[26];
        }
    }
    /** Initialize your data structure here. */
    TrieNode root;
    public Trie() {
       root = new TrieNode();
        
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {   //O(n)
        TrieNode n= root;
        for(int i=0; i<word.length();i++){
            if(n.children[word.charAt(i)-'a'] ==null){
                n.children[word.charAt(i)-'a'] = new TrieNode();
            }
            n= n.children[word.charAt(i)-'a'];
        }
        n.isEnd= true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {   // O(n)
        TrieNode n= root;
        for(int i=0; i<word.length();i++){
            if(n.children[word.charAt(i)-'a'] ==null){
                return false;
            }
            n= n.children[word.charAt(i)-'a'];
        }
        return n.isEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {  //O(n)
        TrieNode n= root;
        for(int i=0; i<prefix.length();i++){
            if(n.children[prefix.charAt(i)-'a'] ==null){
                return false;
            }
            n= n.children[prefix.charAt(i)-'a'];
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */